From 4f33c7508c4e1c1643ad2588f18050587412e564 Mon Sep 17 00:00:00 2001
From: ShenJing <jingshex@intel.com>
Date: Thu, 30 Nov 2017 11:18:34 +0800
Subject: [PATCH 1473/2345] skl: Disable core in skl_shutdown and force dsp
 cores off

This patch is ported from:

skl: force dsp cores off without checking usage_count

The call chain is skl_shutdown=>skl_dsp_sleep=>skl_dsp_put_core.
skl_dsp_put_core checks --skl>cores.usage_count[core_id]. When
the latter is equal to 0, driver puts dsp cores to sleep.

The new issue happens as sometimes, but very rarely, that
--skl>cores.usage_count[core_id] is not equal to 0. dsp core
is on when system goes to shutdown and triggers this new issue.

During reboot testing, it's very easy to hit golbal reset.
PMC reports timeout waiting for IP SIDE_POK_STS and PRIM_POK_STS
deassertion and kernel shutdown successfully.

Debug shows skl_shutdown might be not clean. PMC might wait
skl to be power gated and such wait timeout in the end.

skl_shutdown need put skl dsp quiescent. The patch fixes it by
disabling skl dsp cores.

Change-Id: I33722adb06bdf805cf66bd1e87b3b29efa2a7801
Signed-off-by: Sun, Yi J <yi.j.sun@intel.com>
Signed-off-by: he, bo <bo.he@intel.com>
Signed-off-by: Bai, Jie A <jie.a.bai@intel.com>
Signed-off-by: ShenJing <jingshex@intel.com>
---
 sound/hda/ext/hdac_ext_stream.c       | 1 -
 sound/soc/intel/skylake/skl-sst-dsp.c | 1 +
 sound/soc/intel/skylake/skl.c         | 3 +++
 3 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/sound/hda/ext/hdac_ext_stream.c b/sound/hda/ext/hdac_ext_stream.c
index f0906dd..f8ac962 100644
--- a/sound/hda/ext/hdac_ext_stream.c
+++ b/sound/hda/ext/hdac_ext_stream.c
@@ -504,7 +504,6 @@ void snd_hdac_ext_stop_streams(struct hdac_ext_bus *ebus)
 	if (bus->chip_init) {
 		list_for_each_entry(stream, &bus->stream_list, list)
 			snd_hdac_stream_stop(stream);
-		snd_hdac_bus_stop_chip(bus);
 	}
 }
 EXPORT_SYMBOL_GPL(snd_hdac_ext_stop_streams);
diff --git a/sound/soc/intel/skylake/skl-sst-dsp.c b/sound/soc/intel/skylake/skl-sst-dsp.c
index 19ee1d4..886e524 100644
--- a/sound/soc/intel/skylake/skl-sst-dsp.c
+++ b/sound/soc/intel/skylake/skl-sst-dsp.c
@@ -278,6 +278,7 @@ int skl_dsp_disable_core(struct sst_dsp *ctx, unsigned int core_mask)
 
 	return ret;
 }
+EXPORT_SYMBOL(skl_dsp_disable_core);
 
 int skl_dsp_boot(struct sst_dsp *ctx)
 {
diff --git a/sound/soc/intel/skylake/skl.c b/sound/soc/intel/skylake/skl.c
index 82b0053..4830d9d 100644
--- a/sound/soc/intel/skylake/skl.c
+++ b/sound/soc/intel/skylake/skl.c
@@ -968,6 +968,9 @@ static void skl_shutdown(struct pci_dev *pci)
 		return;
 
 	snd_hdac_ext_stop_streams(ebus);
+	snd_hdac_ext_bus_link_power_down_all(ebus);
+	skl_dsp_sleep(skl->skl_sst->dsp);
+	skl_dsp_disable_core(skl->skl_sst->dsp, SKL_DSP_CORE0_ID);
 	list_for_each_entry(s, &bus->stream_list, list) {
 		stream = stream_to_hdac_ext_stream(s);
 		snd_hdac_ext_stream_decouple(ebus, stream, false);
-- 
2.7.4

